frontend/ui/widget/textviewer.lua

```diff
@@ -112,17 +119,67 @@ function TextViewer:init()
         show_parent = self,
     }
 
-    local buttons = self.buttons_table or
+    local default_buttons =
         {
             {
-                {
-                    text = _("Close"),
-                    callback = function()
-                        self:onClose()
-                    end,
-                },
+                text = _("Close"),
+                callback = function()
+                    self:onClose()
+                end,
+                hold_callback = self.default_hold_callback,
+            },
+            {
+                text = _("Find"),
+                callback = function()
+                    local input_dialog
+                    input_dialog = InputDialog:new{
+                        title = _("Enter text to search for"),
+                        input = self.search_value,
+                        buttons = {
+                            {
+                                {
+                                    text = _("Cancel"),
+                                    callback = function()
+                                        UIManager:close(input_dialog)
+                                    end,
+                                },
+                                {
+                                    text = _("Find first"),
+                                    callback = function()
+                                        self:findCallback(input_dialog, true)
+                                    end,
+                                },
+                                {
+                                    text = _("Find next"),
+                                    is_enter_default = true,
+                                    callback = function()
+                                        self:findCallback(input_dialog)
+                                    end,
+                                },
+                            },
+                        },
+                    }
+
+                    self.check_button_case = CheckButton:new{
+                        text = _("Case sensitive"),
+                        checked = self.case_sensitive,
+                        parent = input_dialog,
+                        callback = function()
+                            self.case_sensitive = self.check_button_case.checked
+                        end,
+                    }
+                    input_dialog:addWidget(self.check_button_case)
+
+                    UIManager:show(input_dialog)
+                    input_dialog:onShowKeyboard()
+                end,
+                hold_callback = self.default_hold_callback,
             },
         }
+    local buttons = self.buttons_table or { default_buttons }
+    if self.add_default_buttons then
+        table.insert(buttons, default_buttons)
```

Comment:
If no buttons_table, and add_default_buttons=true, won't default_buttons be added twice ?

Reply:
Yes, the caller shouldn't do that. Need a comment?

Reply:
https://github.com/koreader/koreader/blob/b40878479e0d0b3f175bcede0fac30610fd545fb/frontend/ui/widget/textviewer.lua#L63

Reply:
Can't you just take care of it?
```lua
    local buttons = self.buttons_table or {}
    if self.add_default_buttons or not self.buttons_table then
        table.insert(buttons, default_buttons)
    end
```